mantidproject/mantid
# New algorithm CorrectTOFAxis

This PR fixes partially the issue #18091 by introducing a new algorithm `CorrectTOFAxis` which can be used to adjust the TOF axis of a workspace so that conversion from 'TOF' to 'DeltaE' will result in the elastic peak being at the zero energy transfer. The complete fix will require the removal of similar adjustments in [`LoadILLTOF`](http://docs.mantidproject.org/nightly/algorithms/LoadILLTOF-v1.html), and the usage of this algorithm together with [`GetEiMonDet`](http://docs.mantidproject.org/nightly/algorithms/GetEiMonDet-v2.html) to get the correct TOF axis for ILL workspaces. `CorrectTOFAxis` in itself is instrument/institution independent, however.

The unit tests contain some functionality common to `GetEiMonDet2Test`. This functionality was moved to WorkspaceCreationHelpers and `GetEiMonDet2Test` was refactored accordingly.

**To test:**

The script below exemplifies the algorithm's three modes of operation. Using all the modes should give the same results: shift the TOF axis slightly and update the 'Ei' and 'wavelength' sample logs. The converted workspaces should have the elastic peaks at the zero energy transfer (although in this case the raw workspace comes pretty close as well).

```python
# Adjust the input path to the Mantid build directory
inputFile = '<mantid-build-directory>/ExternalData/Testing/Data/UnitTest/ILL/IN4/084446.nxs'
raw = Load(inputFile)
# Calibrate incident energy.
EPP = FindEPP(raw)
newEi = GetEiMonDet(raw, EPP, '1-299', 0, PulseInterval=3000)
# Correct the TOF axis directly... (updates the 'Ei' and 'wavelength' sample logs)
correctedWithEPPTable = CorrectTOFAxis(raw, EPPTable=EPP, ReferenceSpectra='1-299', EFixed=newEi)
# ...by using elastic bin index...
elasticChannel = raw.run().getProperty('monitor.elasticpeak').value
correctedWithBinIndex = CorrectTOFAxis(raw, ReferenceSpectra='1-299', ElasticBinIndex=elasticChannel, EFixed=newEi)
# ... or use a reference workspace. (also updates the sample logs)
referencedRaw = CorrectTOFAxis(raw, ReferenceWorkspace=correctedWithBinIndex)
# Convert the units
convertedCorrected = ConvertUnits(correctedWithEPPTable, Target='DeltaE', EMode='Direct')
convertedRaw= ConvertUnits(raw, Target='DeltaE', EMode='Direct')
```


*The release notes were updated accordingly.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
